extensions/amp-ooyala-player/0.1/amp-ooyala-player.js

```diff
@@ -0,0 +1,89 @@
+/**
+ * Copyright 2016 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {isLayoutSizeDefined} from '../../../src/layout';
+import {loadPromise} from '../../../src/event-helper';
+import {user} from '../../../src/log';
+
+class AmpOoyalaPlayer extends AMP.BaseElement {
+
+  /** @override */
+  preconnectCallback(onLayout) {
+    this.preconnect.url('https://player.ooyala.com', onLayout);
+  }
+
+  /** @override */
+  createPlaceholderCallback() {
+    const placeholderUrl = this.element.getAttribute('data-placeholder');
```

Comment:
What @jridgewell said. This callback is only for generating dynamic placeholders. Users should already be able to do this  without you overriding this callback. Drop the entire callback.

If your service has a simple endpoint to retrieve thumbnails or poster for a video (e.g `https://thumbs.ooyala.com/${this.videoId_}/${this.otherParam_}/${this.width_}x${this.height_}`) you can then use this as your dynamic placeholder if the user haven't provided one using the [normal placeholder mechanism that AMP supports](https://github.com/ampproject/amphtml/blob/master/spec/amp-html-layout.md#placeholder).

See amp-instagram for an example usage.